### ISSUE TITLE
Pagination not responsive on mobile 

### ISSUE THREAD

**Issue by blueline86** [joomla/joomla-cms#32674]
### Steps to reproduce the issue
Create enough content items to have ~10 or more pages of items in list view on the front end. 

View the front end on a mobile device. 

### Expected result
The width of the pagination never exceeds 100vw when there are lots of pages. Ideally some items in the middle should be removed and replaced with an ellipsis to maintain a maximum width within the viewport. Eg: First,1,2,3,...,8,9,10,Last. 

### Actual result
The pagination is wider than the rest of the view and creates extra space off screen to the right of the content. When scrolled the window can move left to right due to the overflow. 

### System information (as much as possible)
Php 7.4
Joomla 4 beta 7

**Comment by ghost** [comment]
I want to contribute to this.

**Comment by Bond97** [comment]
It also happens in Atum template (Joomla4 backend).

![Screenshot_20210505-013813_Vivaldi](https://user-images.githubusercontent.com/78050514/117079898-62fc9080-ad45-11eb-834d-8cc2c6019c39.jpg)

**Comment by morphinestyle** [comment]
I'm also waiting for a fix for the mobile...
Better reduce to 6 max element to show, or customizable element number by backend.
I hope it will solved as soon as possible: it's strange that this issue is still present.

**Comment by morphinestyle** [comment]
Here a fix for fontend J4 by pagination override.

1. make an override of joomla/pagination layout
2. edit YOUR-TEMPLATE/html/layouts/joomla/pagination/list.php
3. after `<?php echo $list['previous']['data']; ?>` add this code:

```
<?php
$maxPages = 5;

$list["pages"] = array_values($list["pages"]);
while (count($list["pages"]) > $maxPages) {
    foreach ($list["pages"] as $key => $val) {
        if (!$val["active"]) {
            break;
        }
    }
    if ($key > $maxPages / 2) {
        array_shift($list["pages"]);
    } else {
        array_pop($list["pages"]);
    }
}
?>
```

4. change maxPages value to set the max number of page displayed by pagination

**Comment by coolcat-creations** [comment]
@richard67 can you change the flag to backend template as its not a frontend screen?

**Comment by richard67** [comment]
@coolcat-creations The issue was reported for the frontend according to its description, that's why it was added by @chmst . A later comment said it also affects backend. Will remove the label for now.

**Comment by coolcat-creations** [comment]
Ah thank you, just submitted a PR for Frontend so if that will be hopefully merged it will only be a backend issue

**Comment by richard67** [comment]
Please test #37711 for the frontend. Thanks in advance.

As the comment above https://github.com/joomla/joomla-cms/issues/32674#issuecomment-832301576 says, there is also an issue in backend, so I leave this here open.

**Comment by pl71** [comment]
Nice hack, but I've expected also some ellipsis inside.
Could be extended to adjust the length according to the viewport?<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/32674">issues.joomla.org/tracker/joomla-cms/32674</a>.</sub>

**Comment by morphinestyle** [comment]
No, it's simple PHP and it cannot detect viewport size.

**Comment by blueline86** [comment]
The solution may be best handled with JavaScript so the viewport can be taken into consideration?

**Comment by morphinestyle** [comment]
Yes, javascript can help to make it responsive/adaptive by viewport.

**Comment by brianteeman** [comment]
it does not need javascript. it can be done with pure css.

Two examples

https://codepen.io/brianteeman/pen/ExLapJV

https://codepen.io/brianteeman/pen/dyePjrN

But the first is probably the best

Maybe this is something that @crystalenka can look at with the JXT

**Comment by Kadmon06** [comment]
Until the changes get implemented, as I didn't really know where to add the code and the css, I made a slider for the navigation. In the blog.php I've included `<div style="overflow-x:auto;">` before the navigation, and it solved the mobile-friendliness for me.

**Comment by danielmarschall** [comment]
Is there any update on this? This bug was reported nearly 2 years ago and it impacts the mobile friendliness out-of-the-box...

**Comment by morphinestyle** [comment]
Use my fix: https://github.com/joomla/joomla-cms/issues/32674#issuecomment-1095065752

**Comment by richard67** [comment]
Reopening as the PR has been closed.

**Comment by hans2103** [comment]
I think we are looking for something like the video below. It is done by a template override on joomla.pagination.list.
I will try to find out how this can be pushed back into the project on a proper way.  

https://github.com/joomla/joomla-cms/assets/639822/a97f64bb-e849-42d5-9166-8e147c0c0efe

**Comment by hans2103** [comment]
My example was not about the arrows. My example is about hiding elements when there is space to place them and prevent the pagination go out of bounds.

Arrows, chevrons and/or text for the outer pagination links it out of scope for the issue.

**Comment by morphinestyle** [comment]
Of course @hans2103, i know this issue well, i fixed it temporarily with an override to reduce the page numbers by PHP with https://github.com/joomla/joomla-cms/issues/32674#issuecomment-1095065752

**Comment by hans2103** [comment]
Work in progress https://github.com/hans2103/joomla-cms/tree/feature/5.0-dev/responsive-pagination

I am adding a `page-item-offset-[#]` className to the list items of the pagination (except the list items containing className `.start`, `.prev`, `.active`, `.next`, `.last`)
The page items are initially hidden with display none and will be shown after certain breakpoints.

The `...` have not been added yet. 

feel free to attribute https://github.com/hans2103/joomla-cms/commit/c2da14ce1c192d81f0f26f1f0b56a13a539ba39f

**Comment by brianteeman** [comment]
> I am adding a `page-item-offset-[#]` className to the list items of the pagination (except the list items containing className `.start`, `.prev`, `.active`, `.next`, `.last`) The page items are initially hidden with display none and will be shown after certain breakpoints.

No need to add anything. Surely this is the ideal usecase for the css `nth-of-type` etc

Something as simple as this scss should do it. 
```

.page-item {
  @media (max-width: 767.98px) {
    display: none;

    &:nth-of-type(4) {
      position: relative;
      padding-right: 50px;

      &::after {
        content: '\2026';
        position: absolute;
        font-size: 24px;
        top: 0;
        left: 45px;
      }
    }

    &:nth-child(-n+4),
    &:nth-last-child(-n+4) {
      display: block;
    }

    &:nth-last-child(-n+4) {
    	&::after {
    		content: none;
    	}
    }
  }
}
```

Which will produce 
![image](https://github.com/joomla/joomla-cms/assets/1296369/c43c78d8-125b-47e9-b57e-7bc7d89b59e8)

Just add it to 
build\media_source\templates\site\cassiopeia\scss\vendor\bootstrap\_pagination.scss
build\media_source\templates\administrator\atum\scss\vendor\bootstrap\_pagination.scss

**Comment by hans2103** [comment]
@brianteeman please do not use `px` but `rem` instead. 
People might change the default font-size of their browser or other tools and your suggestion does not follow this change.

**Comment by brianteeman** [comment]
It's not a pr just an example for those better than me to follow as a start. It's the same I posted as a codepen last year

**Comment by polovsek** [comment]
> @richard67 can you change the flag to backend template as its not a frontend screen?

For backend template you make override \administrator\templates\YOUT_BACKEND_TEMPLATE\html\layouts\joomla\pagination\links.php and
before <?php foreach ($pages['pages'] as $k => $page) : ?> you add:

<?php
$maxPages = 5;

$pages["pages"] = array_values($pages["pages"]);
while (count($pages["pages"]) > $maxPages) {
    foreach ($pages["pages"] as $key => $val) {
        if (!$val["active"]) {
            break;
        }
    }
    if ($key > $maxPages / 2) {
        array_shift($pages["pages"]);
    } else {
        array_pop($pages["pages"]);
    }
}
?>

**Comment by web54** [comment]
Hello,
I encountered the same issue with one site, and discovered this topic.
So I tested in cassiopeia : and the issue is there too.
So, how to enahnce cassiopeia with at least the css code given by @brianteeman ?
it should be an option ( or not )

**Comment by brianteeman** [comment]
Please test #43766

**Comment by richard67** [comment]
Closing as having a pull request. Please test #43766 . Thanks in advance.

**Comment by brianteeman** [comment]
Please re-open

**Comment by pl71** [comment]
I tried to make a JS solution.
It takes the width of the container element in this case, `ul.pagination`.
And it tries to accommodate the active element as well as surrounding items in the space provided.
Still, I'm not sure how to deal with two visible pagination bars (if possible at all).
You can see it in [CodePen](https://codepen.io/pl71/pen/pvyKovV?editors=1111).<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/32674">issues.joomla.org/tracker/joomla-cms/32674</a>.</sub>
